#######################################################################################################################
## Theoretical Tryptic Peptides Part - 1: Helper Function to cleave a single protein
#######################################################################################################################

Tryptic_Peptide_Generator <- function(Protein, Max_Missed_Cleavages = 0){
  
  ## Initialize vector to store tryptic peptides
  Tryptic_Peptides <- character(length = 100000)
  Number_of_Tryptic_Peptides <- 0
  
  ## Protein is a string of amino acids that is non-empty.
  Protein_End <- nchar(Protein)
  
  ## Get Arginine and Lysine Sites
  Arginine_Sites <- unlist(gregexpr(pattern = "R", Protein))
  Lysine_Sites <- unlist(gregexpr(pattern = "K", Protein))
  Proline_Sites <- unlist(gregexpr(pattern = "P", Protein))
  
  ## Remove the "-1" produced by no detection
  Arginine_Sites <- Arginine_Sites[which(Arginine_Sites > 0)]
  Lysine_Sites <- Lysine_Sites[which(Lysine_Sites > 0)]
  Proline_Sites <- Proline_Sites[which(Proline_Sites > 0)]
  
  ## Combine to get all cleavage sites
  cleavage_sites <- c(Arginine_Sites, Lysine_Sites)
  
  ## Sort the cleavage sites in ascending order
  if(length(cleavage_sites) > 0){
    cleavage_sites <- sort(cleavage_sites)
  }
  
  ## Remove cleavage sites that are followed by a Proline
  if(length(cleavage_sites) > 0 & length(Proline_Sites) > 0){
    for(i in 1:length(cleavage_sites)){
      if(is.element((cleavage_sites[i] + 1), Proline_Sites)){
        cleavage_sites <- cleavage_sites[-i]
      }
    }
  }
  
  ## A protein terminating R or K is not a cleavage site
  cleavage_sites <- cleavage_sites[which(cleavage_sites != Protein_End)]
  
  ## vector of locations of starts and ends of substrings ("The Tryptic Peptides")
  substring_positions <- c(0,cleavage_sites, Protein_End)
  
  ## Create tryptic peptides
  for(Missed_Cleavages in 0:Max_Missed_Cleavages){
    
    ## Check whether there are enough cleavage sites to "miss"
    if(length(cleavage_sites) >= Missed_Cleavages){
      
      for(j in 1:(length(substring_positions) - Missed_Cleavages - 1)){
        
        ## Define start and end position for the tryptic peptide in the protein AA sequence
        start_pos <- substring_positions[j] + 1
        end_pos <- substring_positions[j + Missed_Cleavages + 1]
        
        
        
        ## Update the number of tryptic peptides for the protein
        Number_of_Tryptic_Peptides <- Number_of_Tryptic_Peptides + 1
        
        Tryptic_Peptides[Number_of_Tryptic_Peptides] <- str_sub(Protein,
                                                                start = start_pos, 
                                                                end = end_pos)
        
      }
    }
  }
  
  ## Truncate Length
  length(Tryptic_Peptides) <- Number_of_Tryptic_Peptides
  
  ## Empty string issue is fixed
  
  ## DO NOT Remove duplicates
  
  return(Tryptic_Peptides)
}

#######################################################################################################################
## Theoretical Tryptic Peptides Part - 2: Main function to generate database
#######################################################################################################################

## Peptide Database Generator with MC Annotations
Generate_Peptide_Database <- function(Proteome, Maximum_Missed_Cleavages = 0, Minimum_Length = 1, 
                                      Protein_Indices = Selected_Proteins){
  
  print(paste0("Obtaining protein sequences."))
  ## Load Sanquin Proteome for input
  Protein_List <- Proteome
  
  ### Protein_List is the output generated by reading in Uniprot Proteome with read.fasta
  ### Maximum_Missed_Cleavages controls the number of missed cleavage sites 
  ### Minimum length is set to 1 Amino Acid
  
  # Maintain a count of all (non-unique) tryptic peptides generated to truncate list at the end
  count <- 0
  
  Sequence <- character(length = 20000000)
  Protein <-  character(length = 20000000)
  MC <- integer(length = 20000000)
  
  Number_of_Proteins <- length(Protein_List)
  Number_of_Tryptic_Peptides <- 0
  
  print(paste0("Generating tryptic peptides from selected proteins..."))
  
  for(i in Protein_Indices){
    
    tryptic_peptides <- lapply(0:Maximum_Missed_Cleavages,
                               function(x) Tryptic_Peptide_Generator(Protein = Protein_List[i][[1]][1],
                                                                     Max_Missed_Cleavages = x))
    
    ## Initialize vector containing information about the number of missed cleavages generating the peptide
    mc_vec <- c(rep(0, length(tryptic_peptides[[1]])))
    
    ## Add Missed cleavage annotations for non-zero missed cleavage peptides (if they are generated)
    if(Maximum_Missed_Cleavages > 0){
      for(j in 1:Maximum_Missed_Cleavages){
        mc_vec <- c(mc_vec, rep(j, length(tryptic_peptides[[j+1]]) - length(tryptic_peptides[[j]])))
      }
    }
    
    ## Remove Peptides below the minimum length threshold (1 by defaulr)
    mc_vec <- mc_vec[which(nchar(tryptic_peptides[[(Maximum_Missed_Cleavages+1)]]) >= Minimum_Length)]
    tryptic_peptides <- tryptic_peptides[[(Maximum_Missed_Cleavages+1)]][which(
      nchar(tryptic_peptides[[(Maximum_Missed_Cleavages+1)]]) >= Minimum_Length)]
    
    ## Compute the number of tryptic peptides after length thresholding
    Number_of_Tryptic_Peptides <- length(tryptic_peptides)
    
    if(Number_of_Tryptic_Peptides > 0){
      ## Add Protein Label from the Protein List
      Protein[(count+1):(count + Number_of_Tryptic_Peptides)] <- str_replace_all(
        str_extract(attr(Protein_List[i], "name"), "\\|(.+)\\|"), "\\|","")
      
      ## Add Sequences and corresponding missed cleavages
      Sequence[(count+1):(count + Number_of_Tryptic_Peptides)] <- tryptic_peptides 
      MC[(count+1):(count + Number_of_Tryptic_Peptides)] <- mc_vec
    }
    
    ## Update the total number of tryptic peptides that have been generated
    count <- count + Number_of_Tryptic_Peptides
  }
  
  ## Truncate the lengths of all the vectors down to the total count of tryptic peptides
  length(Protein) <- length(Sequence) <- length(MC) <- count
  
  Tryptic_Peptides <- tibble(Peptide = Sequence, Protein, Missed_Cleavage = MC)
  Tryptic_Peptides <- Tryptic_Peptides %>% group_by(Peptide, Protein, Missed_Cleavage) %>% 
                        summarise(Multiplicity = n())
  Tryptic_Peptides <- Tryptic_Peptides %>% dplyr::ungroup()
  Tryptic_Peptides <- Tryptic_Peptides %>% mutate(Length = nchar(Peptide))
  
  return(Tryptic_Peptides)
}


